typing/ctype.ml

```diff
@@ -687,6 +687,12 @@ let check_scope_escape level ty =
         let var = newvar2 level in
         raise (Unify [(ty,ty); (var, var)])
       );
+      begin match ty.desc with
+      | Tconstr (p, _, _) when level < Path.scope p ->
+        let var = newvar2 level in
+        raise (Unify [(ty,ty); (var, var)])
+      | _ -> ()
+      end;
```

Comment:
In `Ctype.update_level` there are several other constructors which have a scope-as-path: `Tpackage`, `Tobject` and `Tvariant`. Is there a reason for handling only Tconstr scopes here? Wouldn't it be nicer to guarantee that if `check_scope_escape` succeeds then `update_level` will not fail?

Reply:
Good question!

I approached that change from another angle, so I didn't even consider it. But, to answer your question: I could handle `Tpackage` in `check_scope_escape` (`Tobject` and `Tvariant` don't need to be handled, because `update_level` will let them through anyway, it will just erase the name), although it's unlikely that the package path would escape the scope there.
Indeed the package path would have needed to be introduced by the pattern, which I don't think can happen at the moment.

There is a follow-up to your question though: is `check_scope_escape` really just cosmetic? In `update_level` when a `Tconstr` escapes its scope, we try to expand it away before failing, while `check_scope_escape` fails directly.
Again, I don't see how we could end up in a situation where an existential would escape its scope, but could actually be expanded away. We would need to have an equation on that existential, but the scope of that equation must necessarily be smaller or equal to the scope of the existential, so again,we'd have a scope escape.

Nevertheless, for everyone's peace of mind (my own included), I think I'll update this PR to:
- try to expand (once) the path when a type constructor escapes its scope
- handle package paths as `update_level` would